HD-3988: range matrix less io

liaud opened the pull request:
This update the API to have more fined grained control and to have more efficient get range that span a single group.

Random tests have been added for the read range cases. The old `Decode` has been split into more modular component.

bert-e commented:
# Hello liaud,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/erasurecode/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

liaud committed:
Rework API to split concern.

This split the read API into three main component:

1. A function to linearize the data fragments. This is used when no repair is needed.
   It doesn't perform any checks and doesn't fallback to a repair.

   Useful for: partial data decoding, fast path where other part were not fetched.

2. A function to repair a payload. This requires the correct number of fragment to
   decode a single range.

   Useful for when errors have occured and a repair is required to perform the
   read.

3. A function to validate header.

   Useful for validating the data beforehand to decide whenever fetching other parts
   is necessary or not.

Tests have been adapted, additional random tests added.

liaud committed:
Adapt RangeMatrix computation to handle more specific cases.

This update the function to allow fetching less parts when
the range only span a single encoding group.

liaud committed:
Rename a few variable to match other code.

liaud committed:
backend: Fix comment indent.

liaud commented on backend.go:
```diff
@@ -800,50 +808,110 @@ type DecodeData struct {
-			p := C.decode_fast(C.int(backend.K), cFrags, C.int(len(frags)),
-				(*C.char)(unsafe.Pointer(&data[blockNr*piecesize*backend.K])),
-				C.uint64_t(piecesize*backend.K), &outlen)
+			p := C.linearize(C.int(backend.K), cDataFrags, C.int(len(fragsIndex)),
```
@benoit-a We talked about this but I don't understand why there is no need for `runtime.KeepAlive` for `frags` . It wasn't here before so I guess it is not necessary, but an explanation would be nice :)

liaud replied:
Added a `runtime.KeepAlive`. Previous implementation didn't have this issue as it had the fallback code in case of failure of which directly used the frags array. This is not the case here.

vietnscality commented on backend.go:
```diff
@@ -800,50 +808,110 @@ type DecodeData struct {
 		// in concurrency without need to lock it access
-		go func(blockNr int) {
-			cFrags := C.makeStrArray(C.int(len(frags)))
+		func(chunkIdx int) {
```
We don't launch goroutines anymore? If it is the case, maybe the `wg` is not necessary?

liaud replied:
A rebase mishap (for debugging). Should be fixed.

vietnscality commented on backend.go:
```diff
@@ -853,38 +921,37 @@ func (backend *Backend) DecodeMatrix(frags [][]byte, piecesize int) (*DecodeData
+	chunkSize := pieceSize + backend.headerSize
+	chunkNr := fragRangeLen / chunkSize
+	if chunkNr*chunkSize != fragRangeLen {
+		chunkNr++
```
there is the same compute block of the nb of chunks above, may be extracting this block into a small function is helpful

liaud replied:
Done.

vietnscality commented on backend.go:
```diff
@@ -961,48 +961,129 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
+ *     p4 [-[* -]-]    When this occurs, we will still query, p4 just to decode
+ *                     the relevant requested range. The decoded buffer
+ *                     will look like [p1 p2 p3 p4(*) p1(*) p2(*) p3 p4].
+ *                     Chunks not marked with a '*' are discarded. Note how
```
if discarded, how can we decode that?

liaud replied:
The `*` are contiguous (in the payload sense) and they are the one that are *not* discarded. Is the confusion coming from the negation ? We could in theory only fetch each `*` precisely.

vietnscality replied:
Actually i missed the "not". But i found this example is more difficult to understand than the examples in the test code in another PR (getrange)

vietnscality commented on backend.go:
```diff
@@ -961,48 +961,129 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
+ *     p4 [-[*]- -]
+ *
+ */
+func (backend *Backend) GetRangeMatrix(startIncl, endIncl, pieceSize, fragSize int) *RangeMatrix {
```
Minor: A test for this fct would secure it

liaud replied:
The function is tested by this PR with randoms tests which rely on `GetRangeMatrix`. Given your comment I also added a more classical test with a few cases (other cases are bothersome to compute and I prefer relying on "fuzzing" the API).

vietnscality commented on backend.go:
```diff
@@ -961,48 +961,129 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
+ *                     the heading and trailing is unecessary and could be
+ *                     discarded in the ideal case.
+ *
+ * Perfect cases occur when the request span a single group (column):
```
IIUC We dont have to load the unecessary p1 in this case since no gaps between p2->p4?

liaud replied:
Yes

liaud committed:
range: Adds an additional test for default cases.

liaud committed:
backend: Adds back goroutine for LinearizeMatrix.

liaud committed:
backend: Factorize chunk calculation into its own helper.

liaud committed:
matrix: Ensure that a reference to frags is kept while linearize tasks are running.

liaud committed:
Fixes crash with asan.

It seems that CGO doesn't like copying aligned packed struct.
This was making asan crashing.

vietnscality commented on backend.go:
```diff
@@ -894,48 +983,129 @@ func (backend *Backend) decodeMatrixSlow(frags [][]byte, piecesize int) (*Decode
+ *      2. When the requested range wraps around fragments all fragments
+ *         are always queried.
+ *
+ * (1) is currently necessary to avoid querying multiple time the same
```
```suggestion
 * (1) is currently necessary to avoid querying multiple times the same
```

vietnscality commented on backend.go:
```diff
@@ -925,12 +925,20 @@ func (backend *Backend) LinearizeMatrix(frags []ValidatedFragment, pieceSize int
 
+	/* Tasks above which call into C.linearize keep a pointer of each
+	   fragment to perform their computation. When all goroutines are in the
+	   ffi call, there is no outstanding reference to frags. This ensure
```
Why we don't keep the ref to cDataFrags or frags in ffi call? may be it is impossible?

liaud replied:
Go doesn't track raw pointer because it is outside of its control. You could compute an arbitrary pointer to a gc object. So when the ffi call is ongoing, there is no longer any references from the PoV of the gc.

liaud replied:
To elaborate a bit more. Keeping a reference is exactly what KeepAlive does. By using the object as an argument.

lamphamsy replied:
+1

vietnscality reviewed (approved):
Great, Thanks @liaud !

lamphamsy commented on backend.go:
```diff
@@ -800,91 +808,172 @@ type DecodeData struct {
+			return nil, errors.New("invalid fragment index")
+		}
+
+		if idx >= backend.K {
```
This check might lead to an unexpected case as the gap check is done after this check.
e.g. for k=2, n=1, the fragments index are: 0, 2 => `lastDataFragIdxExcl=1`

liaud replied:
Just to be sure that we are aligned: In your example, `lastDataFragIdxExcl` is used as an exclusive index. So the final data fragments used will be `0`, which look correct to me.

This function ignores coding fragments and fragments are ordered by their index at this point. So if this check triggers, it means that all data fragments have been iterated on. The gap check is only about data fragments which gets "linearized".

Would a comment help to clarify this ?

lamphamsy commented on backend.go:
```diff
@@ -85,9 +74,9 @@ char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uin
-    for (i = 0; i < k && tocopy > 0; i++) {
-        char *f = get_data_ptr_from_fragment(frags[i]);
-        int fsize = get_fragment_payload_size(frags[i]);
+    for (i = 0; i < inlen && tocopy > 0; i++) {
```
Do we expect that `inlen == k`? If yes, is it necessary a sanity check for this condition?

liaud replied:
Not really, consider the range request that retrieve only a single 32K chunk from a data part. In that case `inlen` will be 1 and K will be 2 in 2 + 1.

From the diff it is not obvious that this is the main change of this function. But that's why DecodeMatrix got split. If everything goes right, you might not have all the data part, so you can only "linearize" what you have.

lamphamsy replied:
ah yes, it makes sense.

lamphamsy commented on backend.go:
```diff
@@ -800,91 +808,172 @@ type DecodeData struct {
+	   all tasks completed. */
+	runtime.KeepAlive(frags)
+
 	// if we got some issues, fallback on "slow" decoding
```
This comment should be updated, right?

liaud replied:
Indeed.

lamphamsy commented on backend.go:
```diff
@@ -800,91 +808,172 @@ type DecodeData struct {
 		vect := make([][]byte, len(frags))
 		for j := 0; j < len(frags); j++ {
-			vect[j] = frags[j][i*cellSize : (i+1)*cellSize]
+			if len(frags[j]) != fragRangeLen {
```
Should we move this check block to be outside of the loop `for i := 0; i < chunkInfo.NrChunk; i++`? It's to avoid to perform same checks on iterating `i`

liaud replied:
Indeed. However I would hope that the compiler can refactor that just fine. I'll extract it nonetheless.

lamphamsy reviewed (approved):
LGTM. There are only small comments.

liaud committed:
backend: Remove outdated comment.

liaud committed:
backend: time -> times.

liaud committed:
backend: Extract frag len check outside the decode loop.

liaud commented:
/approve

bert-e commented:
# Queue out of order



The changeset has received all authorizations to enter the merge queue,
however the merge queue is currently in an incoherent state. In order
to protect this pull request, I have not added the changeset to the
queue.

Look into https://devdocs.scality.net/support/faq/#bert-e for guidance about
what to do next to unlock the situation, or ask for help on the #ci-support
Slack channel.





*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.0`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue HD-3988.

Goodbye liaud.
